File: jenkins-cli.go
package main

import (
	"bytes"
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"time"
)

const (
	JENKINS_URL = "http://your-jenkins-server:8080"
	API_TOKEN   = "your-api-token-here"
	USERNAME    = "your-username"
	JOB_NAME    = "jenkins-demo"
)

type JenkinsClient struct {
	BaseURL  string
	Username string
	Token    string
	Client   *http.Client
}

type BuildInfo struct {
	Number int    `json:"number"`
	URL    string `json:"url"`
}

func customEncrypt(data string) string {
	hash := md5.Sum([]byte(data + "jenkins-secret-key"))
	key := hex.EncodeToString(hash[:])[:16]
	
	result := make([]byte, len(data))
	for i, b := range []byte(data) {
		keyByte := key[i%len(key)]
		result[i] = b ^ keyByte
	}
	
	return hex.EncodeToString(result)
}

func customDecrypt(encrypted string) string {
	data, _ := hex.DecodeString(encrypted)
	hash := md5.Sum([]byte(encrypted + "jenkins-secret-key"))
	key := hex.EncodeToString(hash[:])[:16]
	
	result := make([]byte, len(data))
	for i, b := range data {
		keyByte := key[i%len(key)]
		result[i] = b ^ keyByte
	}
	
	return string(result)
}

func NewJenkinsClient() *JenkinsClient {
	decryptedToken := customDecrypt(API_TOKEN)
	return &JenkinsClient{
		BaseURL:  JENKINS_URL,
		Username: USERNAME,
		Token:    decryptedToken,
		Client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

func (j *JenkinsClient) makeRequest(method, url string, body io.Reader) (*http.Response, error) {
	req, err := http.NewRequest(method, url, body)
	if err != nil {
		return nil, err
	}

	req.SetBasicAuth(j.Username, j.Token)
	req.Header.Set("Content-Type", "application/json")

	return j.Client.Do(req)
}

func (j *JenkinsClient) TriggerJob(jobName string, parameters map[string]string) error {
	var url string
	var body io.Reader

	if len(parameters) > 0 {
		url = fmt.Sprintf("%s/job/%s/buildWithParameters", j.BaseURL, jobName)
		
		params := make(map[string]interface{})
		for k, v := range parameters {
			params[k] = v
		}
		
		jsonData, err := json.Marshal(params)
		if err != nil {
			return fmt.Errorf("failed to marshal parameters: %v", err)
		}
		body = bytes.NewBuffer(jsonData)
	} else {
		url = fmt.Sprintf("%s/job/%s/build", j.BaseURL, jobName)
		body = nil
	}

	resp, err := j.makeRequest("POST", url, body)
	if err != nil {
		return fmt.Errorf("failed to trigger job: connection error")
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusCreated && resp.StatusCode != http.StatusOK {
		bodyBytes, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("jenkins API returned status %d: %s", resp.StatusCode, string(bodyBytes))
	}

	fmt.Printf("✅ Job '%s' triggered successfully\n", jobName)
	
	if location := resp.Header.Get("Location"); location != "" {
		fmt.Printf("📍 Queue URL: %s\n", location)
	}

	return nil
}

func (j *JenkinsClient) GetJobStatus(jobName string) error {
	url := fmt.Sprintf("%s/job/%s/api/json", j.BaseURL, jobName)
	
	resp, err := j.makeRequest("GET", url, nil)
	if err != nil {
		return fmt.Errorf("failed to get job status: connection error")
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("jenkins API returned status %d", resp.StatusCode)
	}

	var jobInfo map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&jobInfo); err != nil {
		return fmt.Errorf("failed to decode response: %v", err)
	}

	fmt.Printf("📊 Job Status for '%s':\n", jobName)
	fmt.Printf("   Name: %v\n", jobInfo["name"])
	fmt.Printf("   Buildable: %v\n", jobInfo["buildable"])
	
	if lastBuild, ok := jobInfo["lastBuild"].(map[string]interface{}); ok && lastBuild != nil {
		fmt.Printf("   Last Build: #%v\n", lastBuild["number"])
		fmt.Printf("   Last Build URL: %v\n", lastBuild["url"])
	}

	return nil
}

func printUsage() {
	fmt.Println("Jenkins CLI Tool")
	fmt.Println("Usage:")
	fmt.Println("  jenkins-cli job [job-name]         - Trigger a Jenkins job")
	fmt.Println("  jenkins-cli status [job-name]      - Get job status")
	fmt.Println("  jenkins-cli help                   - Show this help")
	fmt.Println("")
	fmt.Printf("Default job: %s\n", JOB_NAME)
}

func main() {
	if len(os.Args) < 2 {
		printUsage()
		os.Exit(1)
	}

	client := NewJenkinsClient()
	command := os.Args[1]

	switch command {
	case "job":
		jobName := JOB_NAME
		if len(os.Args) > 2 {
			jobName = os.Args[2]
		}

		fmt.Printf("🚀 Triggering Jenkins job: %s\n", jobName)
		
		// 可以在这里添加参数
		parameters := map[string]string{
			"BRANCH_NAME": "master",
			"BUILD_TYPE": "release",
		}

		if err := client.TriggerJob(jobName, parameters); err != nil {
			fmt.Printf("❌ Error: %v\n", err)
			os.Exit(1)
		}

	case "status":
		jobName := JOB_NAME
		if len(os.Args) > 2 {
			jobName = os.Args[2]
		}

		if err := client.GetJobStatus(jobName); err != nil {
			fmt.Printf("❌ Error: %v\n", err)
			os.Exit(1)
		}

	case "help":
		printUsage()

	default:
		fmt.Printf("❌ Unknown command: %s\n", command)
		printUsage()
		os.Exit(1)
	}
}

File: src/jenkins_cli.rs
use clap::{Arg, Command};
use reqwest::blocking::Client;
use serde_json::{Map, Value};
use std::collections::HashMap;
use std::time::Duration;
use anyhow::{Result, anyhow};
use base64::Engine;

const JENKINS_URL: &str = "http://your-jenkins-server:8080";
const API_TOKEN: &str = "your-api-token-here";
const USERNAME: &str = "your-username";
const JOB_NAME: &str = "jenkins-demo";

struct JenkinsClient {
    base_url: String,
    username: String,
    token: String,
    client: Client,
}

impl JenkinsClient {
    fn new() -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(30))
            .build()
            .expect("Failed to create HTTP client");

        JenkinsClient {
            base_url: JENKINS_URL.to_string(),
            username: USERNAME.to_string(),
            token: API_TOKEN.to_string(),
            client,
        }
    }

    fn get_auth_header(&self) -> String {
        let credentials = format!("{}:{}", self.username, self.token);
        let encoded = base64::engine::general_purpose::STANDARD.encode(credentials.as_bytes());
        format!("Basic {}", encoded)
    }

    fn trigger_job(&self, job_name: &str, parameters: &HashMap<String, String>) -> Result<()> {
        let url = if parameters.is_empty() {
            format!("{}/job/{}/build", self.base_url, job_name)
        } else {
            format!("{}/job/{}/buildWithParameters", self.base_url, job_name)
        };

        let mut request = self.client
            .post(&url)
            .header("Authorization", self.get_auth_header())
            .header("Content-Type", "application/json");

        if !parameters.is_empty() {
            let params: Map<String, Value> = parameters
                .iter()
                .map(|(k, v)| (k.clone(), Value::String(v.clone())))
                .collect();
            request = request.json(&params);
        }

        let response = request
            .send()
            .map_err(|_| anyhow!("connection error"))?;

        if response.status().is_success() {
            println!("✅ Job '{}' triggered successfully", job_name);
            
            if let Some(location) = response.headers().get("Location") {
                if let Ok(location_str) = location.to_str() {
                    println!("📍 Queue URL: {}", location_str);
                }
            }
            Ok(())
        } else {
            let status = response.status();
            let body = response.text().unwrap_or_else(|_| "Unknown error".to_string());
            Err(anyhow!("jenkins API returned status {}: {}", status, body))
        }
    }

    fn get_job_status(&self, job_name: &str) -> Result<()> {
        let url = format!("{}/job/{}/api/json", self.base_url, job_name);

        let response = self.client
            .get(&url)
            .header("Authorization", self.get_auth_header())
            .send()
            .map_err(|_| anyhow!("connection error"))?;

        if response.status().is_success() {
            let job_info: Value = response
                .json()
                .map_err(|_| anyhow!("failed to decode response"))?;

            println!("📊 Job Status for '{}':", job_name);
            
            if let Some(name) = job_info.get("name") {
                println!("   Name: {}", name);
            }
            
            if let Some(buildable) = job_info.get("buildable") {
                println!("   Buildable: {}", buildable);
            }
            
            if let Some(last_build) = job_info.get("lastBuild") {
                if let Some(number) = last_build.get("number") {
                    println!("   Last Build: #{}", number);
                }
                if let Some(url) = last_build.get("url") {
                    println!("   Last Build URL: {}", url);
                }
            }

            Ok(())
        } else {
            Err(anyhow!("jenkins API returned status {}", response.status()))
        }
    }
}

fn print_help() {
    println!("Jenkins CLI Tool");
    println!("Usage:");
    println!("  jenkins-cli job [job-name]         - Trigger a Jenkins job");
    println!("  jenkins-cli status [job-name]      - Get job status");
    println!("  jenkins-cli help                   - Show this help");
    println!();
    println!("Default job: {}", JOB_NAME);
}

fn main() -> Result<()> {
    let matches = Command::new("jenkins-cli")
        .version("1.0.0")
        .about("Jenkins CLI Tool")
        .subcommand(
            Command::new("job")
                .about("Trigger a Jenkins job")
                .arg(
                    Arg::new("job-name")
                        .help("Name of the job to trigger")
                        .value_name("JOB_NAME")
                        .index(1)
                )
        )
        .subcommand(
            Command::new("status")
                .about("Get job status")
                .arg(
                    Arg::new("job-name")
                        .help("Name of the job to check")
                        .value_name("JOB_NAME")
                        .index(1)
                )
        )
        .subcommand(Command::new("help").about("Show help information"))
        .get_matches();

    let client = JenkinsClient::new();

    match matches.subcommand() {
        Some(("job", sub_matches)) => {
            let job_name = sub_matches
                .get_one::<String>("job-name")
                .map(|s| s.as_str())
                .unwrap_or(JOB_NAME);

            println!("🚀 Triggering Jenkins job: {}", job_name);

            // 添加参数
            let mut parameters = HashMap::new();
            parameters.insert("BRANCH_NAME".to_string(), "master".to_string());
            parameters.insert("BUILD_TYPE".to_string(), "release".to_string());

            if let Err(e) = client.trigger_job(job_name, &parameters) {
                eprintln!("❌ Error: {}", e);
                std::process::exit(1);
            }
        }
        
        Some(("status", sub_matches)) => {
            let job_name = sub_matches
                .get_one::<String>("job-name")
                .map(|s| s.as_str())
                .unwrap_or(JOB_NAME);

            if let Err(e) = client.get_job_status(job_name) {
                eprintln!("❌ Error: {}", e);
                std::process::exit(1);
            }
        }
        
        Some(("help", _)) => {
            print_help();
        }
        
        _ => {
            print_help();
            std::process::exit(1);
        }
    }

    Ok(())
}
